File: frontend/twoofus_flutter/lib/models/message.dart
import 'media.dart';

class Message {
  final int id;

  final int senderId;
  final int receiverId;

  final String content;
  final String? nonce;
  final bool isEncrypted;
  final bool isEdited;

  final DateTime createdAt;
  final List<MediaItem> mediaAttachments;

  Message({
    required this.id,
    required this.senderId,
    required this.receiverId,
    required this.content,
    this.nonce,
    this.isEncrypted = false,
    this.isEdited = false,
    required this.createdAt,
    this.mediaAttachments = const [],
  });

  factory Message.fromJson(
    Map<String, dynamic> json,
  ) {
    List<MediaItem> mediaList = [];
    if (json["media"] != null && json["media"] is List) {
      mediaList = (json["media"] as List)
          .map((m) => MediaItem.fromJson(m))
          .toList();
    }

    return Message(
      id: json["id"],

      senderId: json["sender_id"],
      receiverId: json["receiver_id"],

      content: json["content"],
      nonce: json["nonce"],
      isEncrypted: json["is_encrypted"] ?? false,
      isEdited: json["is_edited"] ?? false,

      createdAt: DateTime.parse(
        json["created_at"],
      ),
      mediaAttachments: mediaList,
    );
  }
}

File: backend/app/models/message.py
from typing import Optional, List
from sqlalchemy import Integer, Text, DateTime, ForeignKey, Boolean
from sqlalchemy.orm import Mapped, mapped_column, relationship
from datetime import datetime, timezone

from ..core.database import Base


class Message(Base):
    __tablename__ = "messages"

    id: Mapped[int] = mapped_column(
        Integer,
        primary_key=True,
        index=True
    )

    sender_id: Mapped[int] = mapped_column(
        Integer,
        ForeignKey("users.id")
    )

    receiver_id: Mapped[int] = mapped_column(
        Integer,
        ForeignKey("users.id")
    )

    content: Mapped[str] = mapped_column(Text)

    nonce: Mapped[Optional[str]] = mapped_column(Text, nullable=True)

    is_encrypted: Mapped[bool] = mapped_column(
        Boolean,
        default=False
    )

    is_edited: Mapped[bool] = mapped_column(
        Boolean,
        default=False
    )

    created_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime,
        default=lambda: datetime.now(timezone.utc)
    )

    media_attachments = relationship("Media", back_populates="message", cascade="all, delete-orphan")